node/src/exchange/server.rs

```diff
@@ -1,52 +1,408 @@
+            let response = store
+                .get_by_hash(&hash)
+                .await
+                .map(|head| head.to_header_response())
+                .unwrap_or_else(|_| HeaderResponse::not_found());
+
+            let _ = tx.send((channel, vec![response])).await;
+        });
+    }
+
+    fn handle_request_by_height(&mut self, channel: R::Channel, origin: u64, amount: u64) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+
+        spawn(async move {
+            if !store.has_at(origin).await {
+                let _ = tx.send((channel, vec![HeaderResponse::not_found()])).await;
+                return;
+            }
+
+            let amount = amount.min(MAX_HEADERS_AMOUNT_RESPONSE);
+            let mut responses = Vec::with_capacity(amount as usize);
+
+            for i in origin..origin + amount {
+                match store.get_by_height(i).await {
+                    Ok(h) => responses.push(h.to_header_response()),
+                    Err(_) => break,
+                }
+            }
+
+            if responses.is_empty() {
+                responses.push(HeaderResponse::not_found());
+            }
+
+            let _ = tx.send((channel, responses)).await;
+        });
+    }
+
+    fn handle_invalid_request(&self, channel: R::Channel) {
+        if let Err(TrySendError::Full(response)) =
+            self.tx.try_send((channel, vec![HeaderResponse::invalid()]))
+        {
+            let tx = self.tx.clone();
+
+            spawn(async move {
+                let _ = tx.send(response).await;
+            });
+        }
+    }
+}
+
+fn parse_request(request: HeaderRequest) -> Option<(u64, header_request::Data)> {
+    if !request.is_valid() {
+        return None;
+    }
+
+    let HeaderRequest {
+        amount,
+        data: Some(data),
+    } = request
+    else {
+        return None;
+    };
+
+    Some((amount, data))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::exchange::utils::HeaderRequestExt;
+    use crate::store::tests::gen_filled_store;
+    use celestia_proto::p2p::pb::header_request::Data;
+    use celestia_proto::p2p::pb::{HeaderRequest, StatusCode};
+    use celestia_types::ExtendedHeader;
+    use libp2p::PeerId;
+    use std::future::poll_fn;
+    use std::sync::Arc;
+    use tendermint_proto::Protobuf;
+    use tokio::select;
+    use tokio::sync::oneshot;
+
+    #[tokio::test]
+    async fn request_head_test() {
+        let (store, _) = gen_filled_store(4);
+        let expected_head = store.get_head().unwrap();
+        let mut handler = ExchangeServerHandler::new(Arc::new(store));
+        let (mut sender, receiver) = create_test_response_sender();
+
+        handler.on_request_received(PeerId::random(), "test", HeaderRequest::head_request(), ());
+
+        let received = select! {
+            _ = poll_fn(move |cx| handler.poll(cx, &mut sender)) => panic!("shouldn't return"),
+            r = receiver => { r.unwrap() }
+        };
```

Comment:
Is this needed because of the loop?

Reply:
Yes. Previously, returned `Ready` signaled `poll_fn` to finish, but now, whether is done is opaque from the caller side, so I'm relying on `receiver` to know when to stop.

Reply:
Can you add a helper function that does this? Also add a comment in the function why it is needed.

Reply:
Added `poll_handler_for_result` with explanation at the bottom